File: backend/services/stt.py
"""AssemblyAI speech-to-text service."""
import asyncio
import logging
from typing import Optional

import aiohttp
from aiohttp import TCPConnector

from backend import config

logger = logging.getLogger(__name__)

_UPLOAD_URL = "https://api.assemblyai.com/v2/upload"
_TRANSCRIPT_URL = "https://api.assemblyai.com/v2/transcript"


async def transcribe(audio_bytes: bytes) -> Optional[str]:
    """Upload audio and poll for Japanese transcript. Returns text or None."""
    if not config.ASSEMBLYAI_API_KEY:
        logger.warning("ASSEMBLYAI_API_KEY not configured")
        return None

    headers = {"authorization": config.ASSEMBLYAI_API_KEY}
    connector = TCPConnector(ssl=False)

    try:
        async with aiohttp.ClientSession(connector=connector) as session:
            async with session.post(_UPLOAD_URL, headers=headers, data=audio_bytes) as resp:
                if resp.status != 200:
                    logger.error("Upload failed: %d", resp.status)
                    return None
                audio_url = (await resp.json())["upload_url"]

            async with session.post(
                _TRANSCRIPT_URL,
                headers=headers,
                json={"audio_url": audio_url, "language_code": "ja"},
            ) as resp:
                if resp.status != 200:
                    logger.error("Transcript request failed: %d", resp.status)
                    return None
                transcript_id = (await resp.json())["id"]

            poll_url = f"{_TRANSCRIPT_URL}/{transcript_id}"
            while True:
                async with session.get(poll_url, headers=headers) as resp:
                    if resp.status != 200:
                        logger.error("Polling failed: %d", resp.status)
                        return None
                    data = await resp.json()
                    if data["status"] == "completed":
                        return data["text"]
                    if data["status"] == "error":
                        logger.error("Transcription error: %s", data.get("error"))
                        return None
                await asyncio.sleep(3)
    except Exception as exc:
        logger.error("STT error: %s", exc)
        return None

File: backend/src/services/stt.rs
use std::time::Duration;

use reqwest::Client;
use serde::Deserialize;
use serde_json::json;

use crate::error::{AppError, AppResult};

const UPLOAD_URL: &str = "https://api.assemblyai.com/v2/upload";
const TRANSCRIPT_URL: &str = "https://api.assemblyai.com/v2/transcript";
const POLL_INTERVAL: Duration = Duration::from_millis(750);
const MAX_POLLS: usize = 40; // ~30s

#[derive(Deserialize)]
struct UploadResponse {
    upload_url: String,
}

#[derive(Deserialize)]
struct TranscriptResponse {
    id: String,
    status: String,
    text: Option<String>,
    error: Option<String>,
}

/// Transcribes raw audio bytes (e.g. webm/opus from the browser's
/// MediaRecorder) to Japanese text via AssemblyAI.
#[derive(Clone)]
pub struct SttService {
    http: Client,
    api_key: String,
}

impl SttService {
    pub fn new(http: Client, api_key: String) -> Self {
        Self { http, api_key }
    }

    pub async fn transcribe(&self, audio: Vec<u8>) -> AppResult<Option<String>> {
        let upload: UploadResponse = self
            .http
            .post(UPLOAD_URL)
            .header("authorization", &self.api_key)
            .body(audio)
            .send()
            .await
            .map_err(|e| AppError::Other(anyhow::anyhow!("assemblyai upload failed: {e}")))?
            .json()
            .await
            .map_err(|e| AppError::Other(anyhow::anyhow!("invalid upload response: {e}")))?;

        let create: TranscriptResponse = self
            .http
            .post(TRANSCRIPT_URL)
            .header("authorization", &self.api_key)
            .json(&json!({ "audio_url": upload.upload_url, "language_code": "ja" }))
            .send()
            .await
            .map_err(|e| AppError::Other(anyhow::anyhow!("assemblyai transcript request failed: {e}")))?
            .json()
            .await
            .map_err(|e| AppError::Other(anyhow::anyhow!("invalid transcript response: {e}")))?;

        for _ in 0..MAX_POLLS {
            tokio::time::sleep(POLL_INTERVAL).await;

            let poll: TranscriptResponse = self
                .http
                .get(format!("{TRANSCRIPT_URL}/{}", create.id))
                .header("authorization", &self.api_key)
                .send()
                .await
                .map_err(|e| AppError::Other(anyhow::anyhow!("assemblyai poll failed: {e}")))?
                .json()
                .await
                .map_err(|e| AppError::Other(anyhow::anyhow!("invalid poll response: {e}")))?;

            match poll.status.as_str() {
                "completed" => return Ok(poll.text),
                "error" => {
                    tracing::warn!("assemblyai transcription error: {:?}", poll.error);
                    return Ok(None);
                }
                _ => continue,
            }
        }

        tracing::warn!("assemblyai transcription timed out");
        Ok(None)
    }
}
